src/main/scala/viper/gobra/frontend/Desugar.scala

```diff
@@ -1030,21 +1030,91 @@ object Desugar {
               sendGotPermProxy = methodProxy(BuiltInMemberTag.SendGotPermMethodTag, dchannel.typ, Vector())(src)
             } yield in.Send(dchannel, dmsg, sendChannelProxy, sendGivenPermProxy, sendGotPermProxy)(src)
 
+          case PTypeSwitchStmt(pre, exp, binder, cases, dflt) =>
+            for {
+              dPre <- maybeStmtD(ctx)(pre)(src)
+              dExp <- exprD(ctx)(exp)
+              exprVar <- freshDeclaredExclusiveVar(dExp.typ.withAddressability(Addressability.exclusiveVariable))(dExp.info)
+              exprAss = singleAss(in.Assignee.Var(exprVar), dExp)(dExp.info)
+              _ <- write(exprAss)
+              clauses <- sequence(cases.map(c => typeSwitchCaseD(c, exprVar, binder)(ctx)))
+
+              dfltStmt <- if (dflt.size > 1) {
+                violation("switch statement has more than one default clause")
+              } else if (dflt.size == 1) {
+                stmtD(ctx)(dflt(0))
+              } else {
+                unit(in.Seqn(Vector.empty)(src))
+              }
+
+              clauseBody = clauses.foldRight(dfltStmt){
+                (clauseD, tail) =>
+                  clauseD match {
+                    case (exprCond, body) => in.If(exprCond, body, tail)(body.info)
+                  }
+              }
+            } yield in.Seqn(Vector(dPre, exprAss, clauseBody))(src)
+
           case _ => ???
         }
       }
     }
 
-    def switchCaseD(switchCase: PExprSwitchCase, scrutinee: in.LocalVar)(ctx: FunctionContext): Writer[(in.Expr, in.Stmt)] =
-      switchCase match {
-        case PExprSwitchCase(left, body) => for {
-          acceptExprs <- sequence(left.map(clause => exprD(ctx)(clause)))
-          acceptCond = acceptExprs.foldRight(in.BoolLit(false)(meta(left.head)) : in.Expr){
-            (expr, tail) => in.Or(in.EqCmp(scrutinee, expr)(expr.info), tail)(expr.info)
+    def switchCaseD(switchCase: PExprSwitchCase, scrutinee: in.LocalVar)(ctx: FunctionContext): Writer[(in.Expr, in.Stmt)] = {
+      val left = switchCase.left
+      val body = switchCase.body
+      for {
+        acceptExprs <- sequence(left.map(clause => exprD(ctx)(clause)))
+        acceptCond = acceptExprs.foldRight(in.BoolLit(b = false)(meta(switchCase)): in.Expr){
+          (expr, tail) => in.Or(in.EqCmp(scrutinee, expr)(expr.info), tail)(expr.info)
+        }
+        stmt <- stmtD(ctx)(body)
+      } yield (acceptCond, stmt)
+    }
+
+    def typeSwitchCaseD(switchCase: PTypeSwitchCase, scrutinee: in.LocalVar, bind: Option[PIdnDef])(ctx: FunctionContext): Writer[(in.Expr, in.Stmt)] = {
+      val left = switchCase.left
+      val body = switchCase.body
+      val metaCase = meta(switchCase)
+      for {
+        acceptExprs <- sequence(left.map {
+          case t: PType =>
+            for { e <- exprAndTypeAsExpr(ctx)(t) } yield in.EqCmp(in.TypeOf(scrutinee)(meta(t)), e)(metaCase)
```

Comment:
What if the type is an interface type?

Reply:
I misunderstood Go's language spec and thought that interface types where could not occur in case clauses. That should be fixed in my last commit.